hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java

```diff
@@ -496,12 +505,78 @@ SqlResult execute(ExplainStatementPlan plan) {
+
         JobConfig jobConfig = new JobConfig()
                 .setArgument(SQL_ARGUMENTS_KEY_NAME, args)
                 .setArgument(KEY_SQL_QUERY_TEXT, plan.getQuery())
                 .setArgument(KEY_SQL_UNBOUNDED, plan.isStreaming())
                 .setTimeoutMillis(timeout);
 
+        final Set<Integer> partitions = new HashSet<>();
+        boolean allVariantsValid = true;
+        for (final String mapName : plan.getPartitionStrategyCandidates().keySet()) {
+            var perMapCandidates = plan.getPartitionStrategyCandidates().get(mapName);
+            final var container = getMapContainer(hazelcastInstance.getMap(mapName));
+            final PartitioningStrategy<?> strategy = container.getPartitioningStrategy();
+
+            // We only support Default and Attribute strategies, even if one of the maps uses non-Default/Attribute
+            // strategy, we should abort the process and clear list of already populated partitions so that partition
+            // pruning doesn't get activated at all for this query.
+            if (strategy != null
+                    && !(strategy instanceof DefaultPartitioningStrategy)
+                    && !(strategy instanceof AttributePartitioningStrategy)) {
+                allVariantsValid = false;
+                break;
+            }
+
+            // ordering of attributes matters for partitioning (1,2) produces different partition than (2,1).
+            final List<String> orderedKeyAttributes = new ArrayList<>();
+            if (strategy instanceof AttributePartitioningStrategy) {
+                final var attributeStrategy = (AttributePartitioningStrategy) strategy;
+                orderedKeyAttributes.addAll(asList(attributeStrategy.getPartitioningAttributes()));
+            } else {
+                orderedKeyAttributes.add(DEFAULT_UNIQUE_KEY);
+            }
+
+            for (final Map<String, Expression<?>> perMapCandidate : perMapCandidates) {
+                Object[] partitionKeyComponents = new Object[orderedKeyAttributes.size()];
+                for (int i = 0; i < orderedKeyAttributes.size(); i++) {
+                    final String attribute = orderedKeyAttributes.get(i);
+                    if (!perMapCandidate.containsKey(attribute)) {
+                        // Shouldn't happen, defensive check in case Opt logic breaks and produces variants
+                        // that do not contain all the required partitioning attributes.
+                        throw new HazelcastException("Partition Pruning candidate"
+                                + " does not contain mandatory attribute: " + attribute);
+                    }
+
+                    partitionKeyComponents[i] = perMapCandidate.get(attribute).eval(null, evalContext);
+                }
+
+                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
+                        StrategyUtil.constructKey(partitionKeyComponents)
```

Comment:
this will produce wrong result if the key implements `PartitionAware`, see `DefaultPartitioningStrategy.getPartitionKey`. That is why I said that this code is hard to maintain - it is even hard to get right the first time

Reply:
@ivanthescientist

Reply:
Good catch, I was looking into similar case earlier, but missed that we do in fact support whole OBJECT comparisons in one scenario - when the object is compared to a dynamic param and that can produce candidates. There's no good way to deal with it aside from introducing another PartitioningStrategy that will explicitly ignore PartitionAware, so for now our best bet is to completely disable Pruning optimization for `__key = ?` and `__key = 1`. I added a scenario in this commit to demonstrate (https://github.com/hazelcast/hazelcast/pull/24813/commits/c64493041ee89625b27eaa5274ccd39bd0778290). NonMapped key produces candidates. Mapped one still doesn't of course.

Technically we could probably just disable pruning optimization `_key = ?` for cases where key is of `OBJECT` type in SQL.

Reply:
https://github.com/hazelcast/hazelcast/issues/25033 - issue for the `__key` support. Again this requires essentially ideally introducing another strategy (or extending Attribute one to support whole keys).

Reply:
for DefaultPartitioningStrategy we need to need to use similar logic to DefaultPartitioningStrategy.getPartitionKey. In case of this strategy dynamic param must be of the same type as the IMap key, so we can invoke PartitionAware.getPartitionKey() method on it (the dynamic param!). So something like that:
```java        
    if (strategy instanceof AttributePartitioningStrategy) {
               return PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents)
            } else {
                assert partitionKeyComponents.length == 1;
                return strategy.getPartitionKey(partitionKeyComponents[0]);
            }
```
It is not pretty due to instanceof checks, but we have them also elsewhere.

Reply:
There are the following cases:
1. `AttributePartitioningStrategy` (cannot be used or does not make sense with `ParittionAware`)
  a. query by key attributes - uses `PartitioningStrategyUtil.constructAttributeBasedKey`
  b. query by __key - could use `strategy.getPartitionKey`
2. DefaultPartitioningStrategy
 a.  query by key attributes - not supported, would require constructing key object from attributes
 b. query by __key - can use `strategy.getPartitionKey` regardless if the key is `ParittionAware` or not

Reply:
Or is the problem that in order store expression in `partitionStrategyCandidates` you need to have expression that invokes `strategy.getPartitionKey`? If so, that what we also identified with @Fly-Style for cases where you pass partitioning expression to Jet. Jet will expect single expression that returns a value for which has can be directly calculated by `PartitionService.getPartition`.

Reply:
>We have no way to tell what's happening inside of end user's PartitionAware.getPartitionKey implementation

Don't we? It should obey some contract, which unfortunately is not documented. First reasonable requirement is that it is required to return the same value for the same or equal object. Otherwise entire IMap partitioning is broken, not only SQL.

Reply:
For the record, we decided that:

1. __key member pruning should be supported with the assumption that `PartitionAware` implementation obeys a reasonable contract (like the described above), will be done as separate PR
2. contract for `PartitionAware.getPartitionKey` should be explicitly and strictly documented - yet another PR, also to reviewed by Core team